File: src/taxotag/_tokenizer.py
"""SentencePiece-style unigram tokenizer used by the Gist model."""

from __future__ import annotations

import struct
import unicodedata
from dataclasses import dataclass

_METASPACE = "\u2581"
_CONTROL = (
    set(range(0x01, 0x09))
    | {0x0B}
    | set(range(0x0E, 0x20))
    | {
        0x7F,
        0x8F,
        0x9F,
    }
)
_SPACES = {
    0x09,
    0x0A,
    0x0C,
    0x0D,
    0xA0,
    0x1680,
    0x2028,
    0x2029,
    0x202F,
    0x205F,
    0x2581,
    0x3000,
    0xFEFF,
    0xFFFD,
} | set(range(0x2000, 0x2010))


@dataclass(frozen=True)
class Token:
    """A token ID and the normalized Unicode scalars it covers."""

    id: int
    scalars: tuple[str, ...]


def nmt_normalize(text: str) -> str:
    """Apply the SentencePiece ``nmt_nfkc`` normalization used by Gist."""

    pre_normalized = "".join(
        " " if ord(character) in _SPACES else character
        for character in text
        if ord(character) not in _CONTROL
    )
    if "\uff5e" not in pre_normalized:
        return unicodedata.normalize("NFKC", pre_normalized)
    return "\uff5e".join(
        unicodedata.normalize("NFKC", segment)
        for segment in pre_normalized.split("\uff5e")
    )


class Tokenizer:
    """Parse and run the compact Gist unigram tokenizer."""

    def __init__(self, data: bytes) -> None:
        if len(data) < 21 or data[:4] != b"GSTK":
            raise ValueError("invalid tokenizer header")

        offset = 5
        unk_id, bos_id, eos_id, count = struct.unpack_from("<4i", data, offset)
        offset += 16
        if count <= 0 or count > (len(data) - offset) // 6:
            raise ValueError("invalid tokenizer vocabulary count")

        score_bytes = count * 4
        if offset + score_bytes > len(data):
            raise ValueError("truncated tokenizer scores")
        scores = list(struct.unpack_from(f"<{count}f", data, offset))
        offset += score_bytes

        length_bytes = count * 2
        if offset + length_bytes > len(data):
            raise ValueError("truncated tokenizer piece lengths")
        lengths = struct.unpack_from(f"<{count}H", data, offset)
        offset += length_bytes

        pieces: dict[str, int] = {}
        maximum_length = 1
        for piece_id, length in enumerate(lengths):
            end = offset + length
            if end > len(data):
                raise ValueError("truncated tokenizer piece")
            piece = data[offset:end].decode("utf-8", errors="replace")
            offset = end
            pieces[piece] = piece_id
            maximum_length = max(maximum_length, len(piece))

        if offset != len(data):
            raise ValueError("unexpected trailing tokenizer data")
        if not 0 <= unk_id < count:
            raise ValueError("tokenizer unknown ID is out of range")

        self.bos_id = bos_id
        self.eos_id = eos_id
        self.unk_id = unk_id
        self._scores = scores
        self._pieces = pieces
        self._max_length = min(maximum_length, 32)
        self._unknown_penalty = min(scores) - 10.0

    def encode(self, text: str) -> list[int]:
        """Return content-subword IDs for ``text``."""

        return [token.id for token in self.tokenize(text)]

    def tokenize(self, text: str) -> list[Token]:
        """Return Viterbi-optimal tokens for ``text``."""

        normalized = nmt_normalize(text)
        squeezed: list[str] = []
        last_was_space = True
        for character in normalized:
            if character == " ":
                if last_was_space:
                    continue
                last_was_space = True
            else:
                last_was_space = False
            squeezed.append(character)
        if squeezed and squeezed[-1] == " ":
            squeezed.pop()

        scalar_text = [
            _METASPACE if character == " " else character for character in squeezed
        ]
        scalar_text.insert(0, _METASPACE)
        length = len(scalar_text)
        if length == 0:
            return []

        best = [-1e18] * (length + 1)
        best[0] = 0.0
        back_positions = [-1] * (length + 1)
        back_ids = [-1] * (length + 1)

        for end in range(1, length + 1):
            start_limit = max(0, end - self._max_length)
            for start in range(start_limit, end):
                piece = "".join(scalar_text[start:end])
                token_id = self._pieces.get(piece)
                if token_id is None:
                    continue
                score = best[start] + float(self._scores[token_id])
                if score > best[end]:
                    best[end] = score
                    back_positions[end] = start
                    back_ids[end] = token_id

            unknown_score = best[end - 1] + self._unknown_penalty
            if unknown_score > best[end]:
                best[end] = unknown_score
                back_positions[end] = end - 1
                back_ids[end] = self.unk_id

        tokens: list[Token] = []
        position = length
        while position > 0:
            start = back_positions[position]
            tokens.append(Token(back_ids[position], tuple(scalar_text[start:position])))
            position = start
        tokens.reverse()
        return tokens

File: tests/reference/swift/Tokenizer.swift
import DesertAnt

/// XLM-R / bge-m3 SentencePiece **Unigram** tokenizer, ported to pure Swift and
/// verified to reproduce the training tokenizer's ids exactly (NFKC normalization,
/// no lowercasing, `▁` metaspace, Viterbi over the vocab with a `min_score − 10`
/// unknown penalty). Backed by a compact `gist_tokenizer.bin` whose vocab is
/// per-script pruned across the 101 languages gist supports. gist uses the
/// content sub-words directly (no `<s>`/`</s>`),
/// which it mean-pools into the semantic embedding.
struct Tokenizer {
    struct Token {
        let id: Int
        let scalars: [Unicode.Scalar]
    }

    let bosID: Int
    let eosID: Int
    let unkID: Int

    private let scores: [Float]
    private let index: [String: Int]
    private let unkPenalty: Double
    private let maxLen: Int

    private static let metaspace: Unicode.Scalar = "\u{2581}"  // ▁

    // sentencepiece `nmt_nfkc` normalization (matching @huggingface/transformers'
    // Precompiled normalizer, which the training tokenizer's charsmap uses): drop
    // control chars, map a fixed set of whitespace-ish chars to a plain space,
    // then NFKC (with the fullwidth-tilde edge case). This matches the reference
    // model on ~98% of posts; the full precompiled-charsmap trie is a GA follow-up.
    private static let control: Set<UInt32> =
        Set(0x01...0x08).union([0x0B]).union(0x0E...0x1F).union([0x7F, 0x8F, 0x9F])
    private static let spaces: Set<UInt32> =
        Set([0x09, 0x0A, 0x0C, 0x0D, 0xA0, 0x1680, 0x2028, 0x2029, 0x202F, 0x205F,
             0x2581, 0x3000, 0xFEFF, 0xFFFD]).union(0x2000...0x200F)

    static func nmtNormalize(_ text: String) -> String {
        var pre = String.UnicodeScalarView()
        for scalar in text.unicodeScalars {
            let c = scalar.value
            if control.contains(c) { continue }
            pre.append(spaces.contains(c) ? " " : scalar)
        }
        let s = String(pre)
        let tilde: Character = "\u{FF5E}"
        guard s.contains(tilde) else { return s.nfkc }
        // Foundation-free split (String.components is Foundation, absent on wasm/Android).
        return s.split(separator: tilde, omittingEmptySubsequences: false)
            .map { String($0).nfkc }
            .joined(separator: String(tilde))
    }

    init?(bytes: [UInt8]) {
        // magic "GSTK" + 1 version byte
        guard bytes.count >= 21, bytes.starts(with: [0x47, 0x53, 0x54, 0x4B]) else { return nil }
        var offset = 5

        func readU16() -> Int? {
            guard offset <= bytes.count - 2 else { return nil }
            defer { offset += 2 }
            return Int(bytes[offset]) | Int(bytes[offset + 1]) << 8
        }
        func readU32() -> UInt32? {
            guard offset <= bytes.count - 4 else { return nil }
            defer { offset += 4 }
            return UInt32(bytes[offset])
                | UInt32(bytes[offset + 1]) << 8
                | UInt32(bytes[offset + 2]) << 16
                | UInt32(bytes[offset + 3]) << 24
        }
        func readInt() -> Int? { readU32().map { Int(Int32(bitPattern: $0)) } }

        guard let unk = readInt(), let bos = readInt(), let eos = readInt(),
              let count = readInt(), count > 0,
              count <= (bytes.count - offset) / 6 else { return nil }

        var parsedScores: [Float] = []
        parsedScores.reserveCapacity(count)
        for _ in 0..<count {
            guard let bits = readU32() else { return nil }
            parsedScores.append(Float(bitPattern: bits))
        }

        var lengths: [Int] = []
        lengths.reserveCapacity(count)
        for _ in 0..<count {
            guard let length = readU16() else { return nil }
            lengths.append(length)
        }

        var parsedIndex = [String: Int](minimumCapacity: count)
        var maximumLength = 1
        for (id, length) in lengths.enumerated() {
            guard length <= bytes.count - offset else { return nil }
            let piece = String(decoding: bytes[offset..<(offset + length)], as: UTF8.self)
            offset += length
            parsedIndex[piece] = id
            maximumLength = max(maximumLength, piece.unicodeScalars.count)
        }
        // `parsedIndex` may be slightly smaller than `count`: Swift `String`
        // keys compare by Unicode canonical equivalence, so a few canonically-
        // equivalent pieces in the 101-language vocab collapse to one key. That
        // is correct here — input is NFKC-normalized before matching, so those
        // pieces are indistinguishable anyway. `offset == bytes.count` is the
        // real integrity check.
        guard offset == bytes.count, (0..<count).contains(unk) else { return nil }

        unkID = unk
        bosID = bos
        eosID = eos
        scores = parsedScores
        index = parsedIndex
        maxLen = min(maximumLength, 32)
        unkPenalty = Double(parsedScores.min() ?? 0) - 10.0
    }

    /// Content sub-word ids for `text`, Viterbi-optimal over the unigram vocab.
    func encode(_ text: String) -> [Int] { tokenize(text).map(\.id) }

    func tokenize(_ text: String) -> [Token] {
        let nfkc = Self.nmtNormalize(text)
        // SentencePiece `remove_extra_whitespaces`: trim + collapse space runs.
        var squeezed = [Unicode.Scalar]()
        squeezed.reserveCapacity(nfkc.unicodeScalars.count)
        var lastWasSpace = true
        for scalar in nfkc.unicodeScalars {
            if scalar == " " {
                if lastWasSpace { continue }
                lastWasSpace = true
            } else {
                lastWasSpace = false
            }
            squeezed.append(scalar)
        }
        if squeezed.last == " " { squeezed.removeLast() }
        let normalized = "\u{2581}" + String(String.UnicodeScalarView(
            squeezed.map { $0 == " " ? "\u{2581}" : $0 }))
        let s = Array(normalized.unicodeScalars)
        let n = s.count
        if n == 0 { return [] }

        var best = [Double](repeating: -1e18, count: n + 1); best[0] = 0
        var backPos = [Int](repeating: -1, count: n + 1)
        var backID = [Int](repeating: -1, count: n + 1)
        for i in 1...n {
            let lo = max(0, i - maxLen)
            for j in lo..<i {
                if let tid = index[String(String.UnicodeScalarView(s[j..<i]))] {
                    let sc = best[j] + Double(scores[tid])
                    if sc > best[i] { best[i] = sc; backPos[i] = j; backID[i] = tid }
                }
            }
            let cand = best[i - 1] + unkPenalty
            if cand > best[i] { best[i] = cand; backPos[i] = i - 1; backID[i] = unkID }
        }

        var tokens: [Token] = []
        var i = n
        while i > 0 {
            let j = backPos[i]
            tokens.append(Token(id: backID[i], scalars: Array(s[j..<i])))
            i = j
        }
        return tokens.reversed()
    }
}
